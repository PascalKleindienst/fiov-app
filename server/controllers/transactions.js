// ========================================
// TRANSACTIONS CONTROLLER=================
// ========================================

// List all transactions of the current user.
module.exports.list_all_transactions = (req, res) => {
    req.app.locals.models.Transactions.findAll({ //findAndCountAll
        where: { user_id: req.user.id },
        include: ['category']
    })
        .then( (transactions) => {
            res.json(transactions);
        })
        .catch((err) => {
            res.send(err);
        });
};

// Create a new transaction for the current user.
module.exports.create_transaction = (req, res) => {
    const { description = '', note = '', amount = 0, createdAt = null } = req.body;

    req.app.locals.models.Transactions
        .create({ description, note, amount, createdAt, user_id: req.user.id })
        .then((transaction) => {
            res.json(transaction.get());
        })
        .catch((err) => {
            res.send(err);
        });
};
