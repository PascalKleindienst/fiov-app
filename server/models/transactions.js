// ========================================
// Transactions Model =====================
// ========================================
module.exports = (sequelize, DataTypes) => {
    const Transactions = sequelize.define('Transactions', {
        description: DataTypes.STRING,
        note: DataTypes.TEXT,
        amount: DataTypes.INTEGER
    }, {});

    Transactions.associate = function(models) {
        Transactions.belongsTo(models.Categories, {
            foreignKey: 'category_id',
            onDelete: 'CASCADE',
            as: 'category'
        });

        Transactions.belongsTo(models.User, {
            foreignKey: 'user_id',
            onDelete: 'CASCADE',
            as: 'user'
        });
    };
    
    return Transactions;
};