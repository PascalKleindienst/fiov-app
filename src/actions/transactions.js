// ==============================================
// TRANSACTIONS ACTIONS =========================
// ==============================================
import ajax from 'fetchival';

// =============== Add Transaction ===============

// =============== Remove Transaction ===============

// =============== Update Transaction ===============

// =============== Set Transactions ===============
export const SET_TRANSACTIONS = 'SET_TRANSACTIONS';

export const setTransactions = () => {
    return (dispatch) => {
        return ajax('/api/transactions', { mode: 'cors', credentials: 'same-origin' }).get().then((transactions) => {
            dispatch({
                type: SET_TRANSACTIONS,
                transactions
            });
        });
    };
};