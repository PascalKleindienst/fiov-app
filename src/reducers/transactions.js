// ==============================================
// TRANSACTIONS REDUCERS ========================
// ==============================================
import { SET_TRANSACTIONS } from '../actions/transactions';

const transactionsReducerDefaultState = [];

export default (state = transactionsReducerDefaultState, action) => {
    switch (action.type) {
    case SET_TRANSACTIONS:
        return action.transactions;
    default:
        return state;
    }
};
