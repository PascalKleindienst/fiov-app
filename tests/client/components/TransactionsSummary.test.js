// ==============================================
// TRANSACTIONSSUMMARY COMPONENT TEST ===========
// ==============================================
import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import Container, { TransactionsSummary } from '../../../src/components/TransactionsSummary';
import transactions from '../fixtures/transactions';

describe('my transaction summary component', () => {
    const mockStore = configureMockStore([thunk]);
    const translate = jest.fn(msg => msg);

    test('should render component with multiple transactions correctly', () => {
        const wrapper = shallow(<TransactionsSummary transactionNotVisibleCount={0} transactionCount={21} transactionsTotal={133742} translate={translate} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render component with 1 transaction correctly', () => {
        const wrapper = shallow(<TransactionsSummary transactionNotVisibleCount={0} transactionCount={1} transactionsTotal={9434} translate={translate} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render component with 1 filtered transaction correctly', () => {
        const wrapper = shallow(<TransactionsSummary transactionNotVisibleCount={1} transactionCount={2} transactionsTotal={9434} translate={translate} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should set props correctly', () => {
        const store = mockStore({
            locale: { languages: ['en'] },
            filters: { text: '', sortBy: '', startDate: undefined, endDate: undefined },
            transactions
        });
        const wrapper = shallow(<Container store={store} />);
        expect(wrapper.props().transactionCount).toBe(3);
        expect(wrapper.props().transactionsTotal).toBe(-105199);
        expect(wrapper.props().transactionNotVisibleCount).toBe(0);
    });
});