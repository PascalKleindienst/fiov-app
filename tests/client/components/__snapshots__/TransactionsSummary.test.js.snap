// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my transaction summary component should render component with 1 filtered transaction correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      transactions.total
      <p>
        transactions.not_visible_filter
      </p>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/add"
      >
        transactions.add_transaction
      </Link>
    </div>
  </div>
</div>
`;

exports[`my transaction summary component should render component with 1 transaction correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      transactions.total
      <p>
        transactions.not_filtered
      </p>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/add"
      >
        transactions.add_transaction
      </Link>
    </div>
  </div>
</div>
`;

exports[`my transaction summary component should render component with multiple transactions correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      transactions.total
      <p>
        transactions.not_filtered
      </p>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/add"
      >
        transactions.add_transaction
      </Link>
    </div>
  </div>
</div>
`;
